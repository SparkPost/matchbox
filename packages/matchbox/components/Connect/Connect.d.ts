export default Connect;
declare function Connect(props: any): any;
declare namespace Connect {
    const displayName: string;
    namespace propTypes {
        const left: any;
        const right: any;
        const children: any;
    }
}
