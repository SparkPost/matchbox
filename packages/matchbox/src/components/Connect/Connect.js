import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styles from './Connect.module.scss';
import classnames from 'classnames';

class Connect extends Component {
  static displayName = 'Connect';

  static propTypes = {
    left: PropTypes.node,
    right: PropTypes.node,
    children: PropTypes.node
  };

  render() {
    const {
      left,
      right,
      children
    } = this.props;

    const leftMarkup = left
      ? <div className={styles.Left}>{left}</div>
      : null;

    const rightMarkup = right
      ? <div className={styles.Right}>{right}</div>
      : null;

    const primaryClasses = classnames(
      styles.Primary,
      left && styles.hasLeft,
      right && styles.hasRight,
    );

    return (
      <div className={styles.Connect}>
        {leftMarkup}
        <div className={primaryClasses}>{children}</div>
        {rightMarkup}
      </div>
    );
  }
}

export default Connect;
