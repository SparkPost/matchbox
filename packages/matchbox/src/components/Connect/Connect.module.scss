@import '../../styles/config';

.Connect {
  display: flex;
}

.Primary {
  flex: 1 0;
  *:focus {
    position: relative;
    z-index: 1;
  }
}

.hasLeft > * {
  border-top-left-radius: 0 !important;
  border-bottom-left-radius: 0 !important;
}

.hasRight > * {
  border-top-right-radius: 0 !important;
  border-bottom-right-radius: 0 !important;
}

.Left, .Right {
  flex: 0 0 auto;
  *:focus {
    position: relative;
    z-index: 1;
  }
}

.Left {
  margin-right: -1px;
  * {
    border-top-right-radius: 0;
    border-bottom-right-radius: 0;
  }
}

.Right {
  margin-left: -1px;
  * {
    border-top-left-radius: 0;
    border-bottom-left-radius: 0;
  }
}
