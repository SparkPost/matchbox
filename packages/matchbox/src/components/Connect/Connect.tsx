import React from 'react';
import { Box } from '../Box';
import { tokens } from '@sparkpost/design-tokens';
import styled from 'styled-components';

const FocusHandler = styled(Box)`
  &:focus-within {
    position: relative;
    z-index: ${tokens.zIndex_default};
  }
`;

export type ConnectProps = {
  children?: React.ReactNode;
  left?: React.ReactNode;
  right?: React.ReactNode;
};

function Connect(props: ConnectProps): JSX.Element {
  const { left, right, children } = props;

  const leftMarkup = left ? (
    <FocusHandler flex="0 0 auto" mr="-1px">
      {left}
    </FocusHandler>
  ) : null;

  const rightMarkup = right ? (
    <FocusHandler flex="0 0 auto" ml="-1px">
      {right}
    </FocusHandler>
  ) : null;

  return (
    <Box display="flex">
      {leftMarkup}
      <FocusHandler flex="1">{children}</FocusHandler>
      {rightMarkup}
    </Box>
  );
}

Connect.displayName = 'Connect';
export default Connect;
