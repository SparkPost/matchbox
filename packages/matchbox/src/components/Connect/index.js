export { default as Connect } from './Connect';
