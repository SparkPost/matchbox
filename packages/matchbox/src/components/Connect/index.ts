export { default as Connect } from './Connect';
export * from './Connect';
