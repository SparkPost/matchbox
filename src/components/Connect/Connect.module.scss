@import '../../styles/config';

.Connect {
  position: relative;
  z-index: z-index(content);
  display: flex;
  flex-direction: row;
}

.Primary {
  flex: 1 0 auto;
}

.hasLeft > * {
  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
}

.hasRight > * {
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
}

.Left, .Right {
  flex: 0 0 auto;
  *:focus {
    position: relative;
    z-index: z-index(content) + 1;
  }
}

.Left {
  margin-right: -1px;

  & > * {
    border-top-right-radius: 0;
    border-bottom-right-radius: 0;
  }
}

.Right {
  margin-left: -1px;
  & > * {
    border-top-left-radius: 0;
    border-bottom-left-radius: 0;
  }
}
