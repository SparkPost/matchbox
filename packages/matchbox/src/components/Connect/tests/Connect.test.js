import React from 'react';
import Connect from '../Connect';
import { shallow } from 'enzyme';

describe('Connect', () => {
  let wrapper;
  let props;
  beforeEach(() => {
    props = {
      left: <button>Button on Left</button>,
      right: <button>Button on Right</button>

    };

    wrapper = shallow(<Connect {...props}><input type="text" /></Connect>);
  });

  it('renders correctly with connected components on left & right', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly with connected components on left', () => {
    wrapper.setProps({ right: null });
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly with connected components on right', () => {
    wrapper.setProps({ left: null });
    expect(wrapper).toMatchSnapshot();
  });

});
