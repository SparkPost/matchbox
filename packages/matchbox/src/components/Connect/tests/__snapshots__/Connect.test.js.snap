// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Connect renders correctly with connected components on left & right 1`] = `
<div
  className="Connect"
>
  <div
    className="Left"
  >
    <button>
      Button on Left
    </button>
  </div>
  <div
    className="Primary hasLeft hasRight"
  >
    <input
      type="text"
    />
  </div>
  <div
    className="Right"
  >
    <button>
      Button on Right
    </button>
  </div>
</div>
`;

exports[`Connect renders correctly with connected components on left 1`] = `
<div
  className="Connect"
>
  <div
    className="Left"
  >
    <button>
      Button on Left
    </button>
  </div>
  <div
    className="Primary hasLeft"
  >
    <input
      type="text"
    />
  </div>
</div>
`;

exports[`Connect renders correctly with connected components on right 1`] = `
<div
  className="Connect"
>
  <div
    className="Primary hasRight"
  >
    <input
      type="text"
    />
  </div>
  <div
    className="Right"
  >
    <button>
      Button on Right
    </button>
  </div>
</div>
`;
